src/compiler/program.ts

```diff
@@ -36,36 +36,353 @@ namespace ts {
+
+                return baseUrlResolveNonRelativeModuleName(suffix, baseUrl, supportedExtensions, compilerOptions, host, traceEnabled);
+            }
+
+            // rootDirs does not contain prefix for candidate - fallthrough to load file from candidate location.
+        }
+        else {
+            if (traceEnabled) {
+                trace(host, Diagnostics.rootDirs_option_is_not_specified_using_0_as_candidate_location, candidate);
+            }
+        }
+
+        const resolvedFileName = loadModuleFromFile(supportedExtensions, candidate, failedLookupLocations, !directoryProbablyExists(getDirectoryPath(candidate), host), host, traceEnabled);
+        return {
+            resolvedModule: resolvedFileName ? { resolvedFileName } : undefined,
+            failedLookupLocations
+        };
+    }
+
+    function baseUrlResolveNonRelativeModuleName(moduleName: string, baseUrl: string, supportedExtensions: string[], compilerOptions: CompilerOptions, host: ModuleResolutionHost, traceEnabled: boolean): ResolvedModuleWithFailedLookupLocations {
+        let longestMatchPrefixLength = -1;
+        let matchedPattern: string;
+        let matchedStar: string;
+
+        const failedLookupLocations: string[] = [];
+        if (compilerOptions.paths) {
+            if (traceEnabled) {
+                trace(host, Diagnostics.paths_option_is_specified_looking_for_a_pattern_to_match_module_name_0, moduleName);
+            }
+
+            for (const key in compilerOptions.paths) {
+                const pattern: string = key;
+                const indexOfStar = pattern.indexOf("*");
+                if (indexOfStar !== -1) {
+                    const prefix = pattern.substr(0, indexOfStar);
+                    const suffix = pattern.substr(indexOfStar + 1);
+                    if (moduleName.length >= prefix.length + suffix.length &&
+                        startsWith(moduleName, prefix) &&
+                        endsWith(moduleName, suffix)) {
+
+                        // use length of prefix as betterness criteria
+                        if (prefix.length > longestMatchPrefixLength) {
+                            longestMatchPrefixLength = prefix.length;
+                            matchedPattern = pattern;
+                            matchedStar = moduleName.substr(prefix.length, moduleName.length - suffix.length);
+                        }
+                    }
+                }
+                else if (pattern === moduleName) {
+                    // pattern was matched as is - no need to seatch further
+                    matchedPattern = pattern;
+                    matchedStar = undefined;
+                    break;
+                }
+            }
         }
+
+        if (matchedPattern) {
+            if (traceEnabled) {
+                trace(host, Diagnostics.Module_name_0_matched_pattern_1, moduleName, matchedPattern);
+            }
+
+            for (const subst of compilerOptions.paths[matchedPattern]) {

```

Comment:
In your proposal example, it shows paths can be defined as the follows:
```
"pattern-1": "substitution" | ["list of substitutions"]
```
However, when `const subst of compilerOptions.paths[matchedPattern]` is ran, if my path's substitution is a string, of will iterate over each character of the string. Does this need to change or perhaps normalize the compiler options?